src/Nethermind/Nethermind.Core/Caching/LruCacheLowObject.cs

```diff
@@ -0,0 +1,249 @@
+        }
+
+        private void Replace(TKey key, TValue value)
+        {
+            if (_leastRecentlyUsed < 0)
+            {
+                ThrowInvalidOperationException();
+            }
+
+            int offset = _leastRecentlyUsed;
+            ref var node = ref _items[offset];
+
+            _cacheMap.Remove(node.Key);
+
+            MoveToMostRecent(ref node, offset);
+            node = new(key, value);
+            _cacheMap.Add(key, offset);
+
+            [DoesNotReturn]
+            static void ThrowInvalidOperationException()
+            {
+                throw new InvalidOperationException(
+                    $"{nameof(LruCache<TKey, TValue>)} called {nameof(Replace)} when empty.");
+            }
+        }
+
+
+        private void MoveToMostRecent(ref LruCacheItem node, int offset)
+        {
+            if (node.Next == offset)
+            {
+                Debug.Assert(_leastRecentlyUsed == offset, "this should only be true for a list with only one node");
+                // Do nothing only one node
+            }
+            else
+            {
+                Remove(ref node, offset);
+                AddMostRecent(ref node, offset);
+            }
+        }
+
+        private void Remove(ref LruCacheItem node, int offset)
+        {
+            Debug.Assert(_leastRecentlyUsed >= 0, "This method shouldn't be called on empty list!");
+            if (node.Next == offset)
+            {
+                Debug.Assert(_leastRecentlyUsed == offset, "this should only be true for a list with only one node");
+                _leastRecentlyUsed = -1;
+            }
+            else
+            {
+                _items[node.Next].Prev = node.Prev;
+                _items[node.Prev].Next = node.Next;
+                if (_leastRecentlyUsed == offset)
+                {
+                    _leastRecentlyUsed = node.Next;
+                }
+            }
+        }
+
+        private void AddMostRecent(ref LruCacheItem node, int offset)
+        {
+            if (_leastRecentlyUsed < 0)
+            {
+                SetFirst(ref node, offset);
+            }
+            else
+            {
+                InsertMostRecent(ref node, offset);
+            }
+        }
+
+        private void InsertMostRecent(ref LruCacheItem newNode, int offset)
+        {
+            newNode.Next = _leastRecentlyUsed;
+            ref var node = ref _items[_leastRecentlyUsed];
+            newNode.Prev = node.Prev;
+            _items[node.Prev].Next = offset;
+            node.Prev = offset;
+        }
+
+        private void SetFirst(ref LruCacheItem newNode, int offset)
+        {
+            newNode.Next = offset;
+            newNode.Prev = offset;
+            _leastRecentlyUsed = offset;
+        }
+
+        private struct LruCacheItem(TKey k, TValue v)
+        {
+            public readonly TKey Key = k;
+            public TValue Value = v;
+            public int Next;
+            public int Prev;
```

Comment:
We could using `Deque` (there is one is solution from a package using array under) removing need for those 2 fields.
Would make tracking indexes in the dictionary a bit harder and we would probably need to expose native indexes of Dequeue (instead of collection indexes - as they would change constantly).

But it is potential huge win on size - 8 bytes less for each LruCacheItem.

Reply:
Although it constantly compacts the array if something is removed from the middle, so probably a bad idea in the end.